agent/recorder-agent.c

```diff
@@ -0,0 +1,419 @@
+};
+
+/* *INDENT-OFF* */
+G_DEFINE_TYPE (HwangsaeRecorderAgent, hwangsae_recorder_agent, G_TYPE_APPLICATION)
+/* *INDENT-ON* */
+
+typedef enum
+{
+  RELAY_METHOD_NONE = 0,
+  RELAY_METHOD_START_STREAMING,
+  RELAY_METHOD_STOP_STREAMING
+} relay_methods;
+
+static void
+hwangsae_recorder_agent_send_rest_api (HwangsaeRecorderAgent * self,
+    relay_methods method, gchar * edge_id)
+{
+  CURL *curl;
+  CURLcode res;
+  g_autofree gchar *host =
+      g_settings_get_string (self->settings, "relay-address");
+  guint port = g_settings_get_uint (self->settings, "relay-api-port");
+  g_autofree gchar *url = NULL;
+  g_autofree gchar *postData = NULL;
+
+  struct curl_slist *hs = NULL;
+  hs = curl_slist_append (hs, "Content-Type: application/json");
+
+  if (method == RELAY_METHOD_START_STREAMING) {
+    url =
+        g_strdup_printf ("http://%s:%d/api/v1.0/srt/%s/%s", host, port, "start",
+        edge_id);
+    postData = g_strdup ("{}");
+  } else if (method == RELAY_METHOD_STOP_STREAMING) {
+    url =
+        g_strdup_printf ("http://%s:%d/api/v1.0/srt/%s/%s", host, port, "stop",
+        edge_id);
+    postData = g_strdup ("{}");
+  }
+
+  g_debug ("calling api %s", url);
+
+  curl = curl_easy_init ();
+  if (curl) {
+    curl_easy_setopt (curl, CURLOPT_URL, url);
+    curl_easy_setopt (curl, CURLOPT_HTTPHEADER, hs);
+    if (postData)
+      curl_easy_setopt (curl, CURLOPT_POSTFIELDS, postData);
+    res = curl_easy_perform (curl);
+    curl_easy_cleanup (curl);
+  }
+  return;
+}
+
+static GstClockTime
+get_file_duration (const gchar * file_path)
+{
+  g_autoptr (GstDiscoverer) discoverer = NULL;
+  g_autoptr (GError) error = NULL;
+  g_autoptr (GstDiscovererInfo) info = NULL;
+  g_autoptr (GstDiscovererStreamInfo) stream_info = NULL;
+  g_autoptr (GstCaps) stream_caps = NULL;
+  g_autofree gchar *stream_caps_str = NULL;
+  g_autofree gchar *uri = NULL;
+  const gchar *container_type;
+
+  discoverer = gst_discoverer_new (5 * GST_SECOND, &error);
+  g_assert_no_error (error);
+
+  uri = g_strdup_printf ("file://%s", file_path);
+
+  info = gst_discoverer_discover_uri (discoverer, uri, &error);
+  g_assert_no_error (error);
+  g_assert_cmpint (gst_discoverer_info_get_result (info), ==,
+      GST_DISCOVERER_OK);
+
+  stream_info = gst_discoverer_info_get_stream_info (info);
+  stream_caps = gst_discoverer_stream_info_get_caps (stream_info);
+
+  stream_caps_str = gst_caps_to_string (stream_caps);
+  g_debug ("Container file has caps: %s", stream_caps_str);
+
+  g_assert_cmpint (gst_caps_get_size (stream_caps), ==, 1);
+
+  if (g_str_has_suffix (file_path, ".mp4")) {
+    container_type = "video/quicktime";
+  } else if (g_str_has_suffix (file_path, ".ts")) {
+    container_type = "video/mpegts";
+  } else {
+    g_assert_not_reached ();
+  }
+
+  g_assert_cmpstr
+      (gst_structure_get_name (gst_caps_get_structure (stream_caps, 0)), ==,
+      container_type);
+
+  return gst_discoverer_info_get_duration (info);
+}
```

Comment:
This comes from the test case code and is only for testing. Pls remove.

Reply:
Sure.